cmd.sh

```diff
@@ -57,7 +57,7 @@ cd /opt/nodemcu-firmware
 # EXTRA_CCFLAGS="-DBUILD_DATE=... AND -DNODE_VERSION=..." to make turned into an escaping/expanding nightmare for which
 # I never found a good solution
 if [ "$CAN_MODIFY_VERSION" = true ]; then
-  MODULES=$(awk '/^#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -)
+  MODULES=$(awk '/^[ \t]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -)
```

Comment:
How about using the POSIX `[:space:]` for that like in the line below.

Reply:
Awk doesn't seem to like `[:space:]`...I'll have a look to see why!

Reply:
gawk(1) seems to work, but that means adding it to the list of packages to apt-get in the Dockerfile.  Let me know which you prefer (ie. leaving it as `[ \t]`  using awk, or changing it to `[:space:]` and adding gawk).

Reply:
The shorthand equivalent for `[:space:]` is `\s` and my macOS is happy with that.

```
~/Data/NodeMCU/nodemcu-firmware (dev) > awk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tls,tmr,uart,wifi
```
It's a detail but `\s*` is IMO preferable to `[ \t]*` because it matches all whitespace and not just spaces and tabs (however unlikely it is to find those).

Reply:
With `LUA_USE_MODULES_TCS34725` enabled on my Mac:
```
preston:nodemcu-firmware dimitrit$ awk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tls,tmr,uart,wifi
preston:nodemcu-firmware dimitrit$ awk '/^[[:space:]]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
```
Which implies that `\s` and `[:space:]` are not actually equivalent in macOS.

However, in Ubuntu they do appear to behave the same:
```
preston:nodemcu-firmware dimitrit$ docker run --rm -ti -v `pwd`:/opt/nodemcu-firmware docker-nodemcu-build /bin/bash
root@b9860b50cfe4:/opt/nodemcu-firmware# awk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
root@b9860b50cfe4:/opt/nodemcu-firmware# awk '/^[[:space:]]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
root@b9860b50cfe4:/opt/nodemcu-firmware# exit
```

@marcelstoer, please could you check the above on your Mac? In any case, given that the docker image is based on Ubuntu I agree that it makes sense to change the regex to `[:space:]`.

Reply:
Just did a few more tests...it turns out that in the Ubuntu Docker image `awk` is a symbolic link, which (finally) links to `mawk`:

```
root@d717ed48657b:/opt/nodemcu-firmware# ls -l /usr/bin/awk
lrwxrwxrwx 1 root root 21 Feb 28 19:13 /usr/bin/awk -> /etc/alternatives/awk
root@d717ed48657b:/opt/nodemcu-firmware# ls -l /etc/alternatives/awk
lrwxrwxrwx 1 root root 13 Feb 28 19:13 /etc/alternatives/awk -> /usr/bin/mawk
root@d717ed48657b:/opt/nodemcu-firmware# file /usr/bin/mawk
/usr/bin/mawk: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.24, BuildID[sha1]=4e43393facf439efde422d29048db342528974ad, stripped
```

However, when `gawk` is installed,  `awk` is linked to `gawk`:
```
root@7beef77f07c1:/opt/nodemcu-firmware# which awk
/usr/bin/awk
root@7beef77f07c1:/opt/nodemcu-firmware# ls -l /usr/bin/awk
lrwxrwxrwx 1 root root 21 Feb 28 19:13 /usr/bin/awk -> /etc/alternatives/awk
root@7beef77f07c1:/opt/nodemcu-firmware# ls -l /etc/alternatives/awk
lrwxrwxrwx 1 root root 13 Apr  7 16:58 /etc/alternatives/awk -> /usr/bin/gawk
root@7beef77f07c1:/opt/nodemcu-firmware# file /usr/bin/gawk
/usr/bin/gawk: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=38028e7fa7d204e30764568640ead5bb665aa125, stripped
```

It appears that `mawk` does not support `[:space:]`, and `\s` does not match leading white space, whereas `gawk` works just fine:
```
root@fdcdb3f4c8df:/opt/nodemcu-firmware# /usr/bin/mawk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tls,tmr,uart,wifi
root@fdcdb3f4c8df:/opt/nodemcu-firmware# /usr/bin/mawk '/^[[:space:]]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -

root@fdcdb3f4c8df:/opt/nodemcu-firmware# /usr/bin/gawk '/^\s*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
root@fdcdb3f4c8df:/opt/nodemcu-firmware# /usr/bin/gawk '/^[[:space:]]*#define LUA_USE_MODULES/{print tolower(substr($2,17))}' app/include/user_modules.h | paste -s -d, -
adc,bit,dht,file,gpio,i2c,mqtt,net,node,ow,spi,tcs34725,tls,tmr,uart,wifi
```

It seems to me that the best solution will be to include `gawk` in list of packages to `apt-get` in the Dockerfile; `gawk` does not appear to impact the actual firmware build, and will enable use of POSIX expressions.

Reply:
Woah, all those subtle differences are a real pain; quite a disappointment to be honest. Thanks for testing. I wouldn't want to include `gawk` just to find out that with the next version the behavior changed again. As `\s` (and `[:space:]` for that matter) matches `\r\n\v\f` in addition to space and tab and as those should _really_ never be part of our search space I opt to stick with your original proposal.

Reply:
Agree...those subtleties made this far more painful than it should have been.  

Thanks for merging! :D